Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs

```diff
@@ -48,24 +56,44 @@ public PipGraphFragmentManager(LoggingContext loggingContext, PipExecutionContex
         /// <summary>
         /// Adds a single pip graph fragment to the graph.
         /// </summary>
-        public bool AddFragmentFileToGraph(AbsolutePath filePath, string description)
+        public bool AddFragmentFileToGraph(AbsolutePath filePath, string description, IEnumerable<AbsolutePath> dependencies)
         {
             var deserializer = new PipGraphFragmentSerializer(m_context, new PipGraphFragmentContext());
-            try
+            m_taskMap[filePath] = (deserializer, m_taskFactory.StartNew(async () =>
             {
-                var result = deserializer.Deserialize(
-                    filePath,
-                    (fragmentContext, provenance, pipId, pip) => AddPipToGraph(fragmentContext, provenance, pipId, pip),
-                    description);
-                Logger.Log.DeserializationStatsPipGraphFragment(m_loggingContext, deserializer.FragmentDescription, deserializer.Stats.ToString());
+                var results = await Task.WhenAll(dependencies.Select(dependency => m_taskMap[dependency].Item2));
+                if (results.Any(success => !success))
+                {
+                    return false;
+                }
 
-                return result;
-            }
-            catch (Exception e) when (e is BuildXLException || e is IOException)
-            {
-                Logger.Log.ExceptionOnDeserializingPipGraphFragment(m_loggingContext, filePath.ToString(m_context.PathTable), e.ToString());
-                return false;
-            }
+                try
+                {
+                    var result = await deserializer.DeserializeAsync(
+                        filePath,
+                        m_taskFactory,
+                        (fragmentContext, provenance, pipId, pip) => AddPipToGraph(fragmentContext, provenance, pipId, pip),

```

Comment:
>AddPipToGraph [](start = 69, length = 13)

Capture task to add to pip graph in a list and then do a when all after deserialization (not async) completes.